File: toolbox/bricks/map_manifest.py
"""Read a v2 map manifest — the JSON that replaced `georef.db`.

v2 maps ship a single JSON in the map directory:

    {map_path}/{map_id}_{map_version}_{date}_{time}.json

It is a direct dump of the production objects, which makes it a far better pose
source than the v1 `georef.db`:

| Manifest | Production equivalent |
|---|---|
| `local_origin` — `[lng, lat, alt]` | `GeoRef.local_origin` (same axis order) |
| `geo_levels[]` — `value`, `min_altitude`, `max_altitude` | `GeoLevel` rows |
| `geo_levels[].geo_ref` | the real `geo_ref_id` |
| `geo_keyframes[]` — `x`/`y`/`z`, `orientation` | `GeoKeyframe.position` + ori. |
| `map.venue_type` | `Map.venue_type`, which selects the detection prompts |

**The poses need no conversion.** `georef.db` stores them transposed,
world-to-camera and in WDS/OpenCV, so the v1 path composes three flips to reach
EUS/OpenGL (see `georef_source`). The manifest already stores exactly what
`api_geokeyframe` stores, so that whole class of silent error does not apply here.

## Keyframe identity

The manifest carries **no integer keyframe id** — keyframes are identified by their
image filename (a UUID). We therefore use the **index in `geo_keyframes`** as the
local `video_keyframe_id`, exactly as `api_geokeyframe` ids are opaque integers.

That is safe because *both* halves of the build read the same manifest:
`prepare_runner` maps image filename → index, and `ingest_cli` maps index → pose. They
cannot disagree. The one rule it imposes: **re-exporting the manifest means re-running
prepare and ingest together**, since a different ordering renumbers everything. The
image filename is stored on each `geokeyframe` row so a mismatch is at least visible.
"""

from __future__ import annotations

import json
import re
from dataclasses import dataclass
from pathlib import Path
from urllib.parse import urlparse

import numpy as np

from toolbox.bricks.vendored.geo_transform import Coordinates, GeoTransform, Level
from toolbox.logging import logger

# {map_id}_{map_version}_{date}_{time}.json — e.g. bbhotel-choisy_2_20260805_083206.json
MANIFEST_PATTERN = re.compile(
    r"^(?P<map_id>.+)_(?P<version>\d+)_(?P<stamp>\d{8}_\d{6})\.json$"
)


@dataclass(frozen=True)
class ManifestKeyframe:
    """One keyframe, already in the frames production stores."""

    keyframe_id: int  # index in geo_keyframes
    image_filename: str
    depth_filename: str
    position_eus: np.ndarray  # (3,) float64, metres
    orientation_wxyz: np.ndarray  # (4,) float64, CameraFrame(OpenGL) → LocalFrame(EUS)


@dataclass(frozen=True)
class MapManifest:
    path: Path
    map_id: str
    map_version: int
    map_uuid: str
    venue_type: str | None
    geo_ref_id: int | None
    origin: Coordinates
    levels: tuple[Level, ...]
    keyframes: tuple[ManifestKeyframe, ...]

    def geo_transform(self) -> GeoTransform:
        """EUS↔WGS84 plus the level bands, built exactly as production does."""
        return GeoTransform(origin=self.origin, levels=self.levels)

    def keyframe_poses(self) -> dict[int, ManifestKeyframe]:
        return {kf.keyframe_id: kf for kf in self.keyframes}

    def image_filename_to_keyframe_id(self) -> dict[str, int]:
        return {kf.image_filename: kf.keyframe_id for kf in self.keyframes}

    def depth_filename_by_keyframe_id(self) -> dict[int, str]:
        return {kf.keyframe_id: kf.depth_filename for kf in self.keyframes}


def find_manifest(map_path: Path) -> Path | None:
    """The v2 manifest in a map directory, or None.

    When several exist, the newest wins by the `{date}_{time}` stamp in the name —
    that stamp is why the filename carries it. Ties fall back to mtime.
    """
    map_path = Path(map_path)
    if not map_path.is_dir():
        return None
    candidates = [
        (MANIFEST_PATTERN.match(path.name), path)
        for path in map_path.iterdir()
        if path.is_file() and path.suffix == ".json"
    ]
    matches = [(match, path) for match, path in candidates if match is not None]
    if not matches:
        return None
    if len(matches) > 1:
        logger.warning(
            "%d v2 manifests in %s; using the newest by timestamp. Stale ones renumber "
            "keyframes — remove them to be safe.",
            len(matches),
            map_path,
        )
    matches.sort(key=lambda item: (item[0].group("stamp"), item[1].stat().st_mtime))
    return matches[-1][1]


def _basename(url: str) -> str:
    """Filename an asset URL points at, with any query string discarded.

    Today's manifests point at a public bucket with no query string, so splitting
    on "/" would also work. It is parsed properly because the day a URL arrives
    presigned, the filename would become `abc.jpg?X-Amz-Signature=…` and match
    nothing on disk: no image would resolve to a keyframe id and no depth map
    would be found — both silent. The two other v2 manifest readers
    (`wemap-vision-python`'s `read_v2_map_data`, `wemap-vision-vps-offline`'s
    `download_v2_map_data`) already use `urlparse`, as does `retrieve-map-data`,
    which writes these files; disagreeing with them is the actual risk.
    """
    return Path(urlparse(str(url)).path).name


def load_manifest(path: Path) -> MapManifest:
    """Parse a v2 manifest. Raises `ValueError` on anything structurally missing."""
    path = Path(path)
    data = json.loads(path.read_text(encoding="utf-8"))

    origin_raw = data.get("local_origin")
    if not isinstance(origin_raw, list) or len(origin_raw) != 3:
        raise ValueError(f"{path}: 'local_origin' must be [lng, lat, alt].")
    # Axis order matches Django's PointField: longitude first. Getting this backwards
    # puts the whole map on the wrong continent, so it is spelled out rather than
    # unpacked positionally.
    origin = Coordinates(
        lng=float(origin_raw[0]), lat=float(origin_raw[1]), alt=float(origin_raw[2])
    )

    levels: list[Level] = []
    for entry in data.get("geo_levels") or []:
        levels.append(
            Level(
                value=float(entry["value"]),
                min_altitude=(
                    None
                    if entry.get("min_altitude") is None
                    else float(entry["min_altitude"])
                ),
                max_altitude=(
                    None
                    if entry.get("max_altitude") is None
                    else float(entry["max_altitude"])
                ),
                # Production stores a GEOS geometry; the manifest carries GeoJSON (or
                # null). The vendored GeoTransform tests it with a pure-Python
                # ray-cast, so a dict is what it wants.
                geometry=entry.get("geometry"),
            )
        )
    levels.sort(key=lambda level: level.value)

    geo_ref_ids = {
        int(entry["geo_ref"])
        for entry in data.get("geo_levels") or []
        if entry.get("geo_ref") is not None
    }
    if len(geo_ref_ids) > 1:
        raise ValueError(
            f"{path}: geo_levels reference several geo_refs: {geo_ref_ids}."
        )
    geo_ref_id = next(iter(geo_ref_ids), None)

    keyframes: list[ManifestKeyframe] = []
    for index, entry in enumerate(data.get("geo_keyframes") or []):
        image_filename = _basename(entry.get("image_url", ""))
        if not image_filename:
            raise ValueError(f"{path}: geo_keyframes[{index}] has no image_url.")
        orientation = entry.get("orientation")
        if not isinstance(orientation, list) or len(orientation) != 4:
            raise ValueError(
                f"{path}: geo_keyframes[{index}].orientation must be [w, x, y, z]."
            )
        keyframes.append(
            ManifestKeyframe(
                keyframe_id=index,
                image_filename=image_filename,
                depth_filename=_basename(entry.get("depth_url", "")),
                position_eus=np.array(
                    [float(entry["x"]), float(entry["y"]), float(entry["z"])],
                    dtype=np.float64,
                ),
                orientation_wxyz=np.array(
                    [float(value) for value in orientation], dtype=np.float64
                ),
            )
        )
    if not keyframes:
        raise ValueError(f"{path}: no geo_keyframes.")

    duplicates = len(keyframes) - len({kf.image_filename for kf in keyframes})
    if duplicates:
        raise ValueError(
            f"{path}: {duplicates} keyframe(s) share an image filename, so an image "
            "cannot be mapped to one pose."
        )

    match = MANIFEST_PATTERN.match(path.name)
    map_block = data.get("map") or {}
    manifest = MapManifest(
        path=path,
        map_id=str(
            map_block.get("name") or (match.group("map_id") if match else path.stem)
        ),
        map_version=int(match.group("version")) if match else 0,
        map_uuid=str(map_block.get("uuid") or ""),
        venue_type=map_block.get("venue_type") or None,
        geo_ref_id=geo_ref_id,
        origin=origin,
        levels=tuple(levels),
        keyframes=tuple(keyframes),
    )
    logger.info(
        "Loaded manifest %s: %d keyframes, %d level(s), venue=%s, geo_ref_id=%s",
        path.name,
        len(manifest.keyframes),
        len(manifest.levels),
        manifest.venue_type,
        manifest.geo_ref_id,
    )
    return manifest


def load_map_manifest(map_path: Path) -> MapManifest:
    """Find and load the manifest for a map directory. Raises if there is none."""
    path = find_manifest(Path(map_path))
    if path is None:
        raise FileNotFoundError(
            f"No v2 map manifest in '{map_path}'. Expected a file named "
            "'{map_id}_{version}_{date}_{time}.json'."
        )
    return load_manifest(path)

File: toolbox/backend/src/map-manifest.ts
/**
 * Read a v2 map manifest — the workbench's pose source.
 *
 * This is the TypeScript counterpart of `toolbox/bricks/map_manifest.py`, and the
 * two must agree exactly. Both halves of the build read the same file: the Python
 * side maps image filename → keyframe id when it indexes, this side does the same
 * when it serves. A disagreement would number keyframes differently in the index
 * and in the workbench — wrong image, wrong pose, no error anywhere. So the parse
 * rules, the errors raised, and the basename semantics below are all deliberately
 * copied rather than reinvented.
 *
 * Keyframe ids are **indices into `geo_keyframes`**. The manifest carries no
 * integer id, and production's are opaque anyway. The consequence to keep in mind
 * is that re-exporting a manifest renumbers everything.
 */

import { readFile, readdir, stat } from "node:fs/promises";
import path from "node:path";

/** `{map_id}_{map_version}_{date}_{time}.json` — e.g. `bbhotel-choisy_2_20260805_083206.json`. */
export const MANIFEST_PATTERN = /^(?<mapId>.+)_(?<version>\d+)_(?<stamp>\d{8}_\d{6})\.json$/;

export type ManifestKeyframe = {
  /** Index in `geo_keyframes`. Doubles as the video-keyframe id, as in the bricks. */
  keyframeId: number;
  imageFilename: string;
  depthFilename: string;
  /** EUS metres: +X East, +Y Up, +Z South. */
  positionEus: [number, number, number];
  /** `[w, x, y, z]`, CameraFrame(OpenGL) → LocalFrame(EUS). */
  orientationWxyz: [number, number, number, number];
};

export type ManifestLevel = {
  value: number;
  /**
   * Heights above the map origin, **not** WGS84 altitudes. `null` in the manifest
   * means unbounded, so it becomes ±Infinity here — mapping it to 0 would collapse
   * the band and silently drop every keyframe out of every level.
   */
  minAltitude: number;
  maxAltitude: number;
};

export type MapManifest = {
  path: string;
  mapId: string;
  mapVersion: number;
  mapUuid: string;
  venueType: string | null;
  geoRefId: number | null;
  /** `[lng, lat, alt]` — Django `PointField` order, longitude first. */
  origin: { lng: number; lat: number; alt: number };
  /** Sorted by `value`; first matching band wins, as in the Python. */
  levels: ManifestLevel[];
  keyframes: ManifestKeyframe[];
  keyframeById: Map<number, ManifestKeyframe>;
  keyframeIdByImageFilename: Map<string, number>;
};

/**
 * Filename an asset URL points at, with any fragment and query discarded.
 *
 * Mirrors Python's `Path(urlparse(url).path).name`. Today's manifests point at a
 * public bucket with no query string, so splitting on "/" would also work — but a
 * presigned URL would yield `abc.jpg?X-Amz-Signature=…`, matching nothing on disk:
 * no keyframe id resolved and no depth map found, both silent. `new URL()` is not
 * used because it throws on relative inputs.
 */
export function assetBasename(url: unknown): string {
  const raw = typeof url === "string" ? url : "";
  const withoutFragment = raw.split("#", 1)[0];
  const withoutQuery = withoutFragment.split("?", 1)[0];
  return withoutQuery.slice(withoutQuery.lastIndexOf("/") + 1);
}

/** Level identifier as a string, matching what the bricks emit (`int(value)`). */
export function formatLevel(value: number): string {
  return String(Math.trunc(value));
}

function asFiniteNumber(value: unknown, context: string): number {
  const parsed = typeof value === "number" ? value : Number(value);
  if (!Number.isFinite(parsed)) {
    throw new Error(`${context} must be a finite number, got ${JSON.stringify(value)}.`);
  }
  return parsed;
}

/** `null`/missing altitude bounds are unbounded, as `levels_for_altitudes` treats them. */
function asBound(value: unknown, fallback: number): number {
  if (value === null || value === undefined) {
    return fallback;
  }
  const parsed = Number(value);
  return Number.isFinite(parsed) ? parsed : fallback;
}

export function parseManifest(manifestPath: string, text: string): MapManifest {
  const data = JSON.parse(text) as Record<string, unknown>;
  const label = path.basename(manifestPath);

  const originRaw = data.local_origin;
  if (!Array.isArray(originRaw) || originRaw.length !== 3) {
    throw new Error(`${label}: 'local_origin' must be [lng, lat, alt].`);
  }
  // Longitude first, as in Django's PointField. Getting this backwards puts the
  // whole map on the wrong continent, so it is spelled out rather than unpacked.
  const origin = {
    lng: asFiniteNumber(originRaw[0], `${label}: local_origin[0]`),
    lat: asFiniteNumber(originRaw[1], `${label}: local_origin[1]`),
    alt: asFiniteNumber(originRaw[2], `${label}: local_origin[2]`),
  };

  const levelsRaw = Array.isArray(data.geo_levels) ? data.geo_levels : [];
  const levels: ManifestLevel[] = levelsRaw.map((entry) => {
    const level = (entry ?? {}) as Record<string, unknown>;
    return {
      value: asFiniteNumber(level.value, `${label}: geo_levels[].value`),
      minAltitude: asBound(level.min_altitude, Number.NEGATIVE_INFINITY),
      maxAltitude: asBound(level.max_altitude, Number.POSITIVE_INFINITY),
    };
  });
  levels.sort((a, b) => a.value - b.value);

  const geoRefIds = new Set(
    levelsRaw
      .map((entry) => (entry as Record<string, unknown> | null)?.geo_ref)
      .filter((value) => value !== null && value !== undefined)
      .map((value) => Number(value)),
  );
  if (geoRefIds.size > 1) {
    throw new Error(
      `${label}: geo_levels reference several geo_refs: ${[...geoRefIds].join(", ")}.`,
    );
  }
  const geoRefId = geoRefIds.size === 1 ? [...geoRefIds][0] : null;

  const keyframesRaw = Array.isArray(data.geo_keyframes) ? data.geo_keyframes : [];
  const keyframes: ManifestKeyframe[] = keyframesRaw.map((entry, index) => {
    const keyframe = (entry ?? {}) as Record<string, unknown>;
    const imageFilename = assetBasename(keyframe.image_url);
    if (!imageFilename) {
      throw new Error(`${label}: geo_keyframes[${index}] has no image_url.`);
    }
    const orientation = keyframe.orientation;
    if (!Array.isArray(orientation) || orientation.length !== 4) {
      throw new Error(`${label}: geo_keyframes[${index}].orientation must be [w, x, y, z].`);
    }
    return {
      keyframeId: index,
      imageFilename,
      depthFilename: assetBasename(keyframe.depth_url),
      positionEus: [
        asFiniteNumber(keyframe.x, `${label}: geo_keyframes[${index}].x`),
        asFiniteNumber(keyframe.y, `${label}: geo_keyframes[${index}].y`),
        asFiniteNumber(keyframe.z, `${label}: geo_keyframes[${index}].z`),
      ],
      orientationWxyz: orientation.map((value, axis) =>
        asFiniteNumber(value, `${label}: geo_keyframes[${index}].orientation[${axis}]`),
      ) as [number, number, number, number],
    };
  });
  if (keyframes.length === 0) {
    throw new Error(`${label}: no geo_keyframes.`);
  }

  const keyframeIdByImageFilename = new Map(
    keyframes.map((keyframe) => [keyframe.imageFilename, keyframe.keyframeId]),
  );
  const duplicates = keyframes.length - keyframeIdByImageFilename.size;
  if (duplicates > 0) {
    throw new Error(
      `${label}: ${duplicates} keyframe(s) share an image filename, so an image cannot `
      + "be mapped to one pose.",
    );
  }

  const match = MANIFEST_PATTERN.exec(label);
  const mapBlock = (data.map ?? {}) as Record<string, unknown>;
  return {
    path: manifestPath,
    mapId: String(mapBlock.name || match?.groups?.mapId || path.parse(label).name),
    mapVersion: match ? Number(match.groups?.version) : 0,
    mapUuid: String(mapBlock.uuid ?? ""),
    venueType: mapBlock.venue_type ? String(mapBlock.venue_type) : null,
    geoRefId,
    origin,
    levels,
    keyframes,
    keyframeById: new Map(keyframes.map((keyframe) => [keyframe.keyframeId, keyframe])),
    keyframeIdByImageFilename,
  };
}

/**
 * The v2 manifest in a map directory, or null.
 *
 * When several exist the newest wins by the `{date}_{time}` stamp in the name —
 * that stamp is why the filename carries it. Stale manifests renumber keyframes,
 * so this matches the Python's choice deliberately.
 */
export async function findManifest(mapPath: string): Promise<string | null> {
  let entries: string[];
  try {
    entries = await readdir(mapPath);
  } catch {
    return null;
  }
  const matches = entries
    .map((name) => ({ name, match: MANIFEST_PATTERN.exec(name) }))
    .filter((item): item is { name: string; match: RegExpExecArray } => item.match !== null)
    .sort((a, b) => (a.match.groups!.stamp < b.match.groups!.stamp ? -1 : 1));
  const newest = matches.at(-1);
  return newest ? path.join(mapPath, newest.name) : null;
}

// Keyed on path + mtime + size so an edited manifest is re-read. The promise is
// cached rather than the value: /ui/api/maps polls, and concurrent keyframe
// requests would otherwise each parse the file.
const manifestCache = new Map<string, Promise<MapManifest>>();

/** Load and cache the manifest for a map directory. Throws if there is none. */
export async function loadMapManifest(mapPath: string): Promise<MapManifest> {
  const manifestPath = await findManifest(mapPath);
  if (!manifestPath) {
    throw new Error(
      `No v2 map manifest in '${mapPath}'. Expected a file named `
      + "'{map_id}_{version}_{date}_{time}.json'.",
    );
  }
  const info = await stat(manifestPath);
  const cacheKey = `${manifestPath}:${info.mtimeMs}:${info.size}`;
  const cached = manifestCache.get(cacheKey);
  if (cached) {
    return cached;
  }
  const promise = readFile(manifestPath, "utf8")
    .then((text) => parseManifest(manifestPath, text))
    .catch((error: unknown) => {
      manifestCache.delete(cacheKey);
      throw error;
    });
  manifestCache.set(cacheKey, promise);
  return promise;
}

// ------------------------------------------------------------------ frame adapter

// EUS (East, Up, South) ← WDS (West, Down, South): East = -West, Up = -Down.
const ROT_WDS_TO_EUS = [
  [-1, 0, 0],
  [0, -1, 0],
  [0, 0, 1],
];

// OpenGL (+Y up, -Z forward) ← OpenCV (+Y down, +Z forward).
const ROT_OPENGL_TO_OPENCV = [
  [1, 0, 0],
  [0, -1, 0],
  [0, 0, -1],
];

function matMul3(a: number[][], b: number[][]): number[][] {
  return [0, 1, 2].map((row) =>
    [0, 1, 2].map((col) => a[row][0] * b[0][col] + a[row][1] * b[1][col] + a[row][2] * b[2][col]),
  );
}

/**
 * Rotation matrix for a `[w, x, y, z]` quaternion.
 *
 * Mirrors `toolbox/bricks/vendored/maths/quaternion.py::to_matrix3` — row-major,
 * active rotation. The quaternion is normalised first: `invertRigid4` inverts by
 * transposing the rotation block, which is only the inverse when the block is
 * exactly orthonormal. A quaternion arriving through JSON at 1.0001 in norm would
 * otherwise fold a scale factor into the matrix, and every depth-pinned point
 * would drift in proportion to its distance, with no error raised.
 */
export function quaternionToMatrix3(q: readonly number[]): number[][] {
  const norm = Math.hypot(q[0], q[1], q[2], q[3]);
  if (!Number.isFinite(norm) || norm < 1e-12) {
    throw new Error(`Cannot build a rotation from a degenerate quaternion [${q.join(", ")}].`);
  }
  const [w, x, y, z] = [q[0] / norm, q[1] / norm, q[2] / norm, q[3] / norm];
  return [
    [1 - 2 * (y * y + z * z), 2 * (x * y - w * z), 2 * (x * z + w * y)],
    [2 * (x * y + w * z), 1 - 2 * (x * x + z * z), 2 * (y * z - w * x)],
    [2 * (x * z - w * y), 2 * (y * z + w * x), 1 - 2 * (x * x + y * y)],
  ];
}

/**
 * A manifest keyframe as the 4x4 world-to-camera matrix in WDS this module's
 * consumers expect.
 *
 * Everything downstream in `workbench-index.ts` — headings, the ERP projection,
 * the view cone, `point_world_wds` on the wire — works in the WDS world-to-camera
 * frame that `georef.db` used to store directly. Rather than convert all of it to
 * EUS (and change the meaning of a field the frontend round-trips), this adapter
 * puts a manifest pose back into that frame.
 *
 * It is the bricks' v1 conversion run backwards. Forward, `georef_source` had:
 *
 *     R_eus = ROT_WDS_TO_EUS · R_c2w_wds · ROT_OPENGL_TO_OPENCV
 *     p_eus = ROT_WDS_TO_EUS · t_c2w_wds
 *
 * Both matrices are diagonal ±1, hence self-inverse, so the inverse is the same
 * expression. The Python test suite carries this exact composition as a fixture
 * builder (`_camera_to_world_wds_for_eus`), which is where it was verified.
 */
export function manifestKeyframeToWorldToCameraWds(keyframe: ManifestKeyframe): number[][] {
  const rotationC2w = matMul3(
    matMul3(ROT_WDS_TO_EUS, quaternionToMatrix3(keyframe.orientationWxyz)),
    ROT_OPENGL_TO_OPENCV,
  );
  const [ex, ey, ez] = keyframe.positionEus;
  const translationC2w = [
    ROT_WDS_TO_EUS[0][0] * ex,
    ROT_WDS_TO_EUS[1][1] * ey,
    ROT_WDS_TO_EUS[2][2] * ez,
  ];

  // Invert the rigid transform: R⁻¹ = Rᵀ (safe — the quaternion was normalised).
  const rT = [0, 1, 2].map((row) => [0, 1, 2].map((col) => rotationC2w[col][row]));
  const invT = [0, 1, 2].map(
    (row) =>
      -(
        rT[row][0] * translationC2w[0]
        + rT[row][1] * translationC2w[1]
        + rT[row][2] * translationC2w[2]
      ),
  );
  return [
    [rT[0][0], rT[0][1], rT[0][2], invT[0]],
    [rT[1][0], rT[1][1], rT[1][2], invT[1]],
    [rT[2][0], rT[2][1], rT[2][2], invT[2]],
    [0, 0, 0, 1],
  ];
}
